### PR TITLE
Improvement: DO-1585 custsup ability to add custom legends to causalgraphviewer (1/4)

### PR DESCRIPTION
**Pull request by patricia-causalens**
## Motivation and Context

Make adding custom legend to graphs easier, also expand it to also allow to define nodes in the legend

## Implementation Description

I have updated the type of `additionalLegends` to accept a set of equal props: label, type and symbol which together can be used to define the legend the user wants to show in the graph. It now accepts a list of `GraphLegendDefinition`. 

You can define it as:
```
{
    label: my label
    type: node
    symbol: {
        color: red
    }
} 
```

Depending on the type chosen symbol takes either node properties or edge properties. These are optional and if not set just show the default node/edge color/style. If type is set to none, the user can use that to show a spacer.

Dara side tickets to follow

## Any new dependencies Introduced

None

## How Has This Been Tested?

Locally on storybook

## PR Checklist:

- [x] I have implemented all requirements? (see JIRA, project documentation).
- [x] I am not affecting someone else's work, If I am, they are included as a reviewer.
- [ ] I have added relevant tests (unit, integration or regression).
- [x] I have added comments to all the bits that are hard to follow.
- [x] I have added/updated Documentation.
- [x] I have updated the appropriate changelog with a line for my changes.

## Screenshots (if appropriate):

Screenshot 2024-02-08 at 13 58 25

### CODE REVIEW

**Comment by patricia-causalens** [comment]
Some questions that would be nice to get feedback on:

Is it clearer if I change type none to be type spacer? as that is what people might want to use it for?
Do we think it would be better for extra attributes be within symbol or just at same level as label etc for node and edges?

File: packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx [hunk 1]
```diff
@@ -14,58 +14,98 @@
...
-    { arrowType: 'none', centerSymbol: 'cross', label: 'Prohibited' },
+const ENCODER_LEGEND: GraphLegendDefinition[] = [
+    { label: 'Hard Directed', type: 'edge' },
+    { label: 'Soft Directed', symbol: { arrowType: 'soft' }, type: 'edge' },
+    { label: 'Undirected', symbol: { arrowType: 'none', centerSymbol: 'bidirected' }, type: 'edge' },
+    { label: 'Prohibited', symbol: { arrowType: 'none', centerSymbol: 'cross' }, type: 'edge' },
 ];
 
-const DEFAULT_LEGENDS: Map<EditorMode, LegendLineDefinition[]> = new Map([
+const DEFAULT_LEGENDS: Map<EditorMode, GraphLegendDefinition[]> = new Map([
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx:108]: Should our default legends include the default node color scheme? I believe that was the intention originally but we never got around to doing it

File: packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx [hunk 2]
```diff
@@ -74,8 +114,8 @@ const DEFAULT_LEGENDS: Map<EditorMode, LegendLineDefinition[]> = new Map([
 
 export function getLegendData(
     editorMode: EditorMode,
-    additionalLegend: LegendLineDefinition[]
-): LegendLineDefinition[] {
+    additionalLegend: GraphLegendDefinition[]
+): GraphLegendDefinition[] {
     const modeData = DEFAULT_LEGENDS.get(editorMode) ?? [];
 
     return [...modeData, ...(additionalLegend ?? []).filter(Boolean)];
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx:121]: What's our solution for overriding default legends? As we discussed in the past, this is fine for extra legends but what if the user has a custom color scheme for nodes/edges depending on some property they've defined and they don't want to display the default legend, just their custom one.

One option we've discussed is removing the defaults from here and moving them to the Python side so they can be fully customised

File: packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx [hunk 3]
```diff
@@ -14,58 +14,98 @@
...
+    symbol?: LegendArrowDefinition;
+    type: 'edge';
+}
+
+export interface NodeGraphLegendDefinition extends BaseGraphLegendDefinition {
+    symbol?: LegendNodeDefinition;
+    type: 'node';
+}
+
+export type GraphLegendDefinition = NoneGraphLegendDefinition | EdgeGraphLegendDefinition | NodeGraphLegendDefinition;
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx:63]: This can be simplified to just a discriminated union

```suggestion
export type GraphLegendDefinition = 
    | {
        type: 'none';
    }
    | {
        type: 'edge';
        label: string;
        symbol?: LegendArrowDefinition;
    }
    | {
        type: 'node';
        label: string;
        symbol: LegendNodeDefinition;
    };
```

Then the legend type just becomes:

```tsx
type GraphLegendType = GraphLegendDefinition['type'];
```

File: packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx [hunk 4]
```diff
@@ -14,58 +14,98 @@
...
 
-export interface LegendLineDefinition {
+export interface LegendNodeDefinition {
+    /** defines the filled color of the node symbol */
+    color?: string;
+    /** defines the border color of the node symbol */
+    highlight_color?: string;
+}
+
+export interface LegendArrowDefinition {
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx:26]: What's the reason to rename `Line` to `Arrow`? It's not always an arrow technically, hence the previous 'line' name - sometimes it represents just a straight line. We can just use `Edge` I guess to be consistent with the standard Node/Edge naming

**Review by krzysztof-causalens — CHANGES_REQUESTED** [review]
Looks good, but left a few comments. To address your questions:

> Is it clearer if I change type none to be type spacer? as that is what people might want to use it for?

Agreed `'spacer'` is clearer than `'none'`. It more accurately represents its purpose.

> Do we think it would be better for extra attributes be within symbol or just at same level as label etc for node and edges?

I think we could just make them all flat, I don't really see what we get from nesting the properties.

File: packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx (hunk 1 above)
Comment (tamara-causalens) [reply packages/ui-causal-graph-editor/src/shared/editor-overlay/legend/legend-data.tsx:108]: there should probably be the default there if the user did not pass anything

**Comment by patricia-causalens** [comment]
Example of PAG editor mode with additional legends

Screenshot 2024-02-12 at 19 34 50

**Comment by patricia-causalens** [comment]
@krzysztof-causalens have updated so that all the defaults now live on the python side in Dara. Let me know what you think, I also have a WIP PR there so that you can see how these two relate. There will be 4 PRs total to make sure all dara packages adapt to these changes

**Comment by tamara-causalens** [comment]
> @tamara-causalens have added a default legend for just the node types, let me know if you are happy with the names

Hey Patricia, can you please share a screenshot of what the default looks like, I might be missing it up there? Thanks! The names look good.

**Comment by claudiawestby** [comment]
@patricia-causalens are users able to completely overwrite the legend? like in your pag mode example if I _only_ wanted warning node and success node?

**Comment by patricia-causalens** [comment]
> > @tamara-causalens have added a default legend for just the node types, let me know if you are happy with the names
> 
> Hey Patricia, can you please share a screenshot of what the default looks like, I might be missing it up there? Thanks! The names look good.

This is what the default looks like, for other editor modes it is the same, as they were before but now all of them have the latent,target and other added to them
Screenshot 2024-02-13 at 09 27 22

**Comment by tamara-causalens** [comment]
> > > @tamara-causalens have added a default legend for just the node types, let me know if you are happy with the names
> > 
> > 
> > Hey Patricia, can you please share a screenshot of what the default looks like, I might be missing it up there? Thanks! The names look good.
> 
> This is what the default looks like, for other editor modes it is the same, as they were before but now all of them have the latent,target and other added to them Screenshot 2024-02-13 at 09 27 22

Cool! And given that this is just a plain graph viewer, the arrows are not in the legend as there is only one type?

**Comment by patricia-causalens** [comment]
> > > > @tamara-causalens have added a default legend for just the node types, let me know if you are happy with the names
> > > 
> > > 
> > > Hey Patricia, can you please share a screenshot of what the default looks like, I might be missing it up there? Thanks! The names look good.
> > 
> > 
> > This is what the default looks like, for other editor modes it is the same, as they were before but now all of them have the latent,target and other added to them Screenshot 2024-02-13 at 09 27 22
> 
> Cool! And given that this is just a plain graph viewer, the arrows are not in the legend as there is only one type?

Yes, you can see the PAG mode example above which does have the arrows, but for a default one I added no legends for it as it would just say edge, which didn't seem necessary.

**Comment by patricia-causalens** [comment]
> @patricia-causalens are users able to completely overwrite the legend? like in your pag mode example if I _only_ wanted warning node and success node?

Hi yes, causalgraphviewer now has two legend related props, might be easier to see in the dara side PR. 
They can edit the default_legends prop in which case they can do whatever they want with the default legends for each given editor mode. ALternatively they can add to the default with the additional_legends prop. 

The legends can take an edge related legend, a node related or a spacer.

**Comment by tamara-causalens** [comment]
> > > > > @tamara-causalens have added a default legend for just the node types, let me know if you are happy with the names
> > > > 
> > > > 
> > > > Hey Patricia, can you please share a screenshot of what the default looks like, I might be missing it up there? Thanks! The names look good.
> > > 
> > > 
> > > This is what the default looks like, for other editor modes it is the same, as they were before but now all of them have the latent,target and other added to them Screenshot 2024-02-13 at 09 27 22
> > 
> > 
> > Cool! And given that this is just a plain graph viewer, the arrows are not in the legend as there is only one type?
> 
> Yes, you can see the PAG mode example above which does have the arrows, but for a default one I added no legends for it as it would just say edge, which didn't seem necessary.

Awesome!

File: packages/ui-causal-graph-editor/docs/changelog.md [hunk 5]
```diff
@@ -2,6 +2,10 @@
 title: Changelog
 ---
 
+## NEXT
+
+-   Updated the type of `additionalLegends` to also allow defining nodes
```
Comment (krzysztof-causalens) [inline packages/ui-causal-graph-editor/docs/changelog.md:7]: Let's add a changelog entry for the new `defaultLegends` prop

**Review by krzysztof-causalens — APPROVED** [review]
Looks good, let's just bump up the min legend box width as we discussed offline, since the X button can look a bit too close to the content

File: packages/ui-causal-graph-editor/docs/changelog.md (hunk 5 above)
Comment (patricia-causalens) [reply packages/ui-causal-graph-editor/docs/changelog.md:7]: Added good catch

**Comment by patricia-causalens** [comment]
> Looks good, let's just bump up the min legend box width as we discussed offline, since the X button can look a bit too close to the content

Have added a minHeight and aligned the node to the same space as the center symbol that would appear in an edge.
